File: spring/src/main/java/com/example/demo/security/CustomBCryptPasswordEncoder.java
package com.example.demo.security;

import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class CustomBCryptPasswordEncoder extends BCryptPasswordEncoder {

    private final AuthenticationSecurityService authenticationSecurityService;

    public CustomBCryptPasswordEncoder(AuthenticationSecurityService authenticationSecurityService) {
        this.authenticationSecurityService = authenticationSecurityService;
    }

    @Override
    public String encode(CharSequence rawPassword) {
        return super.encode(rawPassword);
    }

    @Override
    public boolean matches(CharSequence rawPassword, String encodedPassword) {
        try {
            return super.matches(
                    authenticationSecurityService.decrypt(rawPassword.toString()),
                    encodedPassword
            );
        } catch (Exception e) {
            log.error("password decode fail", e);
            e.printStackTrace();

            return false;
        }
    }
}

File: spring/src/main/java/com/example/demo/security/CustomBCryptPasswordEncoder.kt
package com.example.demo.security

import lombok.extern.slf4j.Slf4j
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Component

@Component
@Slf4j
class CustomBCryptPasswordEncoder(private val authenticationSecurityService: AuthenticationSecurityService) : BCryptPasswordEncoder() {
    override fun encode(rawPassword: CharSequence): String {
        return super.encode(rawPassword)
    }

    override fun matches(rawPassword: CharSequence, encodedPassword: String): Boolean {
        return try {
            super.matches(
                    authenticationSecurityService.decrypt(rawPassword.toString()),
                    encodedPassword
            )
        } catch (e: Exception) {
            CustomBCryptPasswordEncoder.log.error("password decode fail", e)
            e.printStackTrace()
            false
        }
    }
}